### PR TITLE
[SPARK-47509][SQL] Block subquery expressions in lambda and higher-order functions

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

This PR blocks subquery expressions in lambda and higher-order functions, since these can generate incorrect results.

### Why are the changes needed?

It is safer to block these query patterns which can generate incorrect results for now. We can consider re-enabling these patterns later after fixing the correctness problem.

### Does this PR introduce _any_ user-facing change?

Yes, see above.

### How was this patch tested?

This PR adds unit test coverage.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by dtenedor** [comment]
cc @cloud-fan

**Review by jchen5 — COMMENTED** [review]
This blocks a subquery that contains a lambda, but the problem case was actually the reverse, a lambda that contains a subquery, e.g. `SELECT arrcol, forall(arrcol, e -> e in (select 1)) as forallout FROM bug_test_table;`

**Comment by jchen5** [comment]
CC @agubichev also

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for making a PR, @dtenedor .

According to the reported JIRA Affected Version, is this Apache Spark 4.0.0 only correctness bug? Do you happen to know when this correctness bug is introduced?

Screenshot 2024-03-21 at 18 22 35

**Comment by dtenedor** [comment]
> According to the reported JIRA Affected Version, is this Apache Spark 4.0.0 only correctness bug? Do you happen to know when this correctness bug is introduced?

@jchen5 do you know?

**Comment by jchen5** [comment]
I suspect it has existed as long as lambdas have. So at least like 5 years?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/higherOrderFunctions.scala [hunk 1]
```diff
@@ -138,4 +141,16 @@ object ResolveLambdaVariables extends Rule[LogicalPlan] {
     case _ =>
       e.mapChildren(resolve(_, parentLambdaMap))
   }
+
+  /**
+   * SPARK-47509: There is a correctness bug when lambdas or higher-order functions appear within
```
Comment (jchen5) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/higherOrderFunctions.scala:146]: Comment should be updated too: appear within -> contain

**Comment by dtenedor** [comment]
@jchen5 @dongjoon-hyun I updated it to switch to checking for subqueries with lambdas inside, instead of vice versa, per request. Please take another look :)

**Review by jchen5 — APPROVED** [review]
Thanks for the fix!

File: common/utils/src/main/resources/error/error-classes.json [hunk 2]
```diff
@@ -4457,6 +4457,11 @@
           "Expressions referencing the outer query are not supported outside of WHERE/HAVING clauses: <sqlExprs>."
         ]
       },
+      "LAMBDA_OR_HIGHER_ORDER_FUNCTION" : {
```
Comment (cloud-fan) [inline common/utils/src/main/resources/error/error-classes.json:4460]: hmmm, are these two the same thing?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/higherOrderFunctions.scala [hunk 3]
```diff
@@ -138,4 +141,16 @@ object ResolveLambdaVariables extends Rule[LogicalPlan] {
...
+
+  /**
+   * SPARK-47509: There is a correctness bug when subquery expressions appear within lambdas or
+   * higher-order functions. Here we check for that case and return an error if the corresponding
+   * configuration indicates to do so.
+   */
+  private def checkForSubqueryExpressions(expression: Expression): Unit = {
+    if (expression.containsPattern(PLAN_EXPRESSION) &&
+      !conf.getConf(SQLConf.ALLOW_SUBQUERY_EXPRESSIONS_IN_LAMBDAS_AND_HIGHER_ORDER_FUNCTIONS)) {
+      throw QueryCompilationErrors.subqueryExpressionInLambdaOrHigherOrderFunctionNotAllowedError()
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/higherOrderFunctions.scala:153]: shall we update the PR title to match the new code?

File: common/utils/src/main/resources/error/error-classes.json (hunk 2 above)
Comment (cloud-fan) [reply common/utils/src/main/resources/error/error-classes.json:4460]: Reading the code comment
```
/**
 * A higher order function takes one or more (lambda) functions and applies these to some objects.
 * The function produces a number of variables which can be consumed by some lambda function.
 */
```

I think for use-facing error message, we can just say "higher order functions". For the code change, we can check both `HigherOrderFunction` and `LambdaFunction` to  be safe.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/higherOrderFunctions.scala (hunk 3 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/higherOrderFunctions.scala:153]: This is done.

File: common/utils/src/main/resources/error/error-classes.json (hunk 2 above)
Comment (dtenedor) [reply common/utils/src/main/resources/error/error-classes.json:4499]: Sounds good, this is simpler. Done.

**Comment by dtenedor** [comment]
Thanks @cloud-fan for your review! Responded to your suggestions and updated the error classes documentation and fixed the conflict with the `master` branch, please take another look.

File: common/utils/src/main/resources/error/error-classes.json [hunk 4]
```diff
@@ -4496,6 +4496,11 @@
...
```
Comment (cloud-fan) [inline common/utils/src/main/resources/error/error-classes.json:4499]:
```suggestion
      "HIGHER_ORDER_FUNCTION" : {
```

File: docs/sql-error-conditions-unsupported-subquery-expression-category-error-class.md [hunk 5]
```diff
@@ -50,6 +50,10 @@ A correlated outer name reference within a subquery expression body was not foun
 
 Expressions referencing the outer query are not supported outside of WHERE/HAVING clauses: `<sqlExprs>`.
 
+## LAMBDA_OR_HIGHER_ORDER_FUNCTION
```
Comment (cloud-fan) [inline docs/sql-error-conditions-unsupported-subquery-expression-category-error-class.md:53]:
```suggestion
## HIGHER_ORDER_FUNCTION
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 6]
```diff
@@ -620,6 +620,12 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase with Compilat
       messageParameters = Map("prettyName" -> toSQLId(prettyName), "syntax" -> toSQLStmt(syntax)))
   }
 
+  def subqueryExpressionInLambdaOrHigherOrderFunctionNotAllowedError(): Throwable = {
+    new AnalysisException(
+      errorClass = "UNSUPPORTED_SUBQUERY_EXPRESSION_CATEGORY.LAMBDA_OR_HIGHER_ORDER_FUNCTION",
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:625]:
```suggestion
      errorClass = "UNSUPPORTED_SUBQUERY_EXPRESSION_CATEGORY.HIGHER_ORDER_FUNCTION",
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveSubquerySuite.scala [hunk 7]
```diff
@@ -270,4 +271,32 @@ class ResolveSubquerySuite extends AnalysisTest {
...
+          child = ScalarSubquery(
+            plan(lv(Symbol("x")))),
+          name = "alias")()
+          :: lv(Symbol("X"))
+          :: Nil))
+
+    assertAnalysisErrorClass(
+      inputPlan = lambdaPlanScanFromTable,
+      expectedErrorClass =
+        "UNSUPPORTED_SUBQUERY_EXPRESSION_CATEGORY.LAMBDA_OR_HIGHER_ORDER_FUNCTION",
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveSubquerySuite.scala:299]:
```suggestion
        "UNSUPPORTED_SUBQUERY_EXPRESSION_CATEGORY.HIGHER_ORDER_FUNCTION",
```

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**